sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/LDeviceAdapter.java

```diff
@@ -359,6 +377,74 @@ protected boolean hasControlBlockCreationCapability(ControlBlockEnum controlBloc
 
     private boolean hasCBNameConf(TServiceSettings tServiceSettings) {
         return tServiceSettings.isSetCbName()
-            && (TServiceSettingsNoDynEnum.CONF.equals(tServiceSettings.getCbName()));
+                && (TServiceSettingsNoDynEnum.CONF.equals(tServiceSettings.getCbName()));
+    }
+
+    /**
+     * Update and/or create Monitoring LNs (LSVS and LGOS) into LDSUIED from ExtRefs binding
+     *
+     * @param tExtRefs           ExtRefs for which source Control Blocks (Goose or SMV) should be monitored
+     * @param tSystemLNGroupEnum LNClass value for LN to monitor (Only LGOS and LSVS are taken into this enum)
+     * @return a list of SclReport Objects that contains errors
+     */
+    public List<SclReportItem> manageMonitoringLns(List<TExtRef> tExtRefs, String doName, TSystemLNGroupEnum tSystemLNGroupEnum) {
+        if (tExtRefs.isEmpty() || (!TSystemLNGroupEnum.LGOS.equals(tSystemLNGroupEnum)
+                && !TSystemLNGroupEnum.LSVS.equals(tSystemLNGroupEnum))) {
+            return Collections.emptyList();
+        }
+        List<SclReportItem> sclReportItems = new ArrayList<>();
+        getLNAdapters().stream().filter(lnAdapter -> tSystemLNGroupEnum.value().equals(lnAdapter.getLNClass())).findFirst()
+                .ifPresentOrElse(lnAdapter -> {
+                    ResumedDataTemplate filter = createDaiFilter(lnAdapter, doName);
+                    List<ResumedDataTemplate> foundDais = lnAdapter.getDAI(filter, true);
+                    if (foundDais.isEmpty()) {
+                        sclReportItems.add(SclReportItem.warning(lnAdapter.getXPath() + "/DOI@name=\"" + doName + "\"/DAI@name=\"setSrcRef\"/Val",
+                                "The DAI cannot be updated"));
+                    } else {
+                        ResumedDataTemplate daToUpdate = foundDais.get(0);
+                        TLN lnToUpdate = lnAdapter.getCurrentElem();
+                        int i = 1;
+                        for (TExtRef tExtRef : tExtRefs) {
+                            if (i != 1) {
+                                this.getCurrentElem().getLN().add(lnToUpdate);
+                            }
+                            updateNewCreatedLnDaiValue(lnToUpdate, tExtRef, String.valueOf(i), daToUpdate).ifPresent(sclReportItems::add);
+                            i++;
+                            lnToUpdate = copySclElement(lnAdapter.getCurrentElem(), TLN.class);
+                        }
```

Comment:
Suggestion : 

- Avoid the extra copy for the last extRef : Do the `copySclElement` at the beginning in the same block as the `getLN().add`.
- Use an old school loop (for int i = 0; ...) to let the loop manage the `i` variable 

```suggestion
                        for (int i = 0; i < tExtRefs.size(); i++) {
                            if (i > 0) {
                                lnToUpdate = copySclElement(lnAdapter.getCurrentElem(), TLN.class);
                                this.getCurrentElem().getLN().add(lnToUpdate);
                            }
                            updateNewCreatedLnDaiValue(lnToUpdate, tExtRefs.get(i), String.valueOf(i + 1), daToUpdate).ifPresent(sclReportItems::add);
                        }
```

Reply:
Your loop start with 0?????
And here you don't copy it each time when i>0 ???
for inst = "1", the LN already exist in LDevice we don't have to add it again.
You should look back your proposition.

Reply:
The proposition pass all the tests.
- For ln.inst, I use `String.valueOf(i + 1)` => when i== 0, inst = 1. You could start the loop at i = 1, but then you have to write `tExtRefs.get(i-1), String.valueOf(i)`. I prefer starting at i=0
- I copy the LN for all extRef, except for the first one (i==0 which means ln.inst == 1). Is it not what you want ?